### DISCUSSION TITLE
Adding process noise in the simulation

### DISCUSSION THREAD

**Discussion by DavidePatria** [do-mpc/do-mpc#122] [Q&A] [answered]
the page relative to [model.v](https://www.do-mpc.com/en/latest/api/do_mpc.model.Model.v.html?highlight=noise#do_mpc.model.Model.v) says in the note

_The measurement noise is used for the do_mpc.estimator.MHE and can be used to simulate a disturbed system in the do_mpc.simulator.Simulator._

but there are no examples showing how to do so.
If I understand correctly it can be used to produce noise in the simulator even if MHE is not used.
I would use it to test the MPC in case of added noise, and having an automatic way to do it would be nice.
thank you

**Comment by 4flixt** [comment, marked as answer]
Hi DavidePatria,

you can just pass values for the process / measurement noise when calling ``simulator.make_step`` as shown [here](https://www.do-mpc.com/en/latest/api/do_mpc.simulator.Simulator.make_step.html).
Typically, you would want to just sample values for the noise at each call of ``simulator_make_step`` from e.g. a normal distribution. 

The code that is repeatedly called should like something like this: 

```python
u0 = mpc.make_step(x0)
w0 = np.random.randn(x0.shape[0],1). # (n,1) vector of sampled noise 
y0 = simulator.make_step(u0,w0=w0)
x0 = estimator.make_step(y0)
```


To allow this you must add process / measurement noise to your model. For (additive) process noise you just need to call ``model.set_rhs`` with the optional argument ``process_noise=True`` ([docs](https://www.do-mpc.com/en/latest/api/do_mpc.model.Model.set_rhs.html)).
And similarly, for measurement noise you have to call ``model.set_meas`` with ``meas_noise =True`` ([docs](https://www.do-mpc.com/en/latest/api/do_mpc.model.Model.set_meas.html)).

**Note**: You have the option to add noise only for selected states / measurements (each call of ``model.set_rhs`` can be configured with / without noise). If not all of your states are subject to process noise you must sample accordingly. 

Hope this helps.

Best
4flixt

**Comment by 4flixt** [comment]
I hope it is ok with you that I converted this to a discussion. This is a new feature and we hope to move some of the issues that are more of a question / discussion to this channel.

**Reply by DavidePatria** [reply]
I found using the issues instrument a bit strange, and it is indeed. This seems a dedicated tool for the purpose, so no problem at all from my side.

**Reply by 4flixt** [reply]
Yes it was long overdue that Github has such a feature. Until now issues were the only option..

**Reply by 4flixt** [reply]
Btw, feel free to mark my reply as an answer to your question. This marks the issue as resolved for me :)

**Comment by DavidePatria** [comment]
If I understand correctly I would have to do 
```v0 = np.random.randn(x0.shape[0],1). # (n,1) vector of sampled noise 
x0 = simulator.make_step(u0,v0=v0)```
and solely the states defined with `meas_noise =True` will be affected, right?

btw this weekend I have done without using thie feature (because I could fully understand) and kinda works.
